src/extensions/windows/Microsoft.DotNet.UpgradeAssistant.Extensions.Windows/WinformsUpdaterSubStep.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions.Windows
+{
+    internal class WinformsUpdaterSubStep : UpgradeStep
+    {
+        private readonly WinformsUpdateStep _winformsUpdateStep;
+        private readonly IUpdater<IProject> _updater;
+
+        public override string Id => _updater.Id;
+
+        public override string Title => _updater.Title;
+
+        public override string Description => _updater.Description;
+
+        public WinformsUpdaterSubStep(WinformsUpdateStep winformsUpdateStep, IUpdater<IProject> updater, ILogger<WinformsUpdateStep> logger)
+            : base(logger)
+        {
+            if (logger is null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            ParentStep = _winformsUpdateStep = winformsUpdateStep ?? throw new ArgumentNullException(nameof(winformsUpdateStep));
+            _updater = updater ?? throw new ArgumentNullException(nameof(updater));
+        }
+
+        protected override async Task<bool> IsApplicableImplAsync(IUpgradeContext context, CancellationToken token)
+        {
+            if (context?.CurrentProject is null)
+            {
+                return false;
+            }
+
+            // Check the updater for an [ApplicableComponents] attribute
+            // If one exists, the step only applies if the project has the indicated components
+            return await context.CurrentProject.IsApplicableAsync(_updater, token).ConfigureAwait(false);
+        }
+
+        protected override async Task<UpgradeStepInitializeResult> InitializeImplAsync(IUpgradeContext context, CancellationToken token)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var currentProject = context.CurrentProject.Required();
+
+            try
+            {
+                var updaterResult = (WinformsUpdaterResult)(await _updater.IsApplicableAsync(context, ImmutableArray<IProject>.Empty.Add(currentProject), token).ConfigureAwait(false));
+                return updaterResult.Result
+                    ? new UpgradeStepInitializeResult(UpgradeStepStatus.Incomplete, $"Winforms updater \"{_updater.Title}\" needs to be applied", _updater.Risk)
+                    : new UpgradeStepInitializeResult(UpgradeStepStatus.Complete, string.Empty, BuildBreakRisk.None);
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
+            catch (Exception exc)
```

Comment:
We're doing a lot of exception handling of generic parameters. Can you create an issue to identify either:

1. Move up the generic handling to higher up
2. Identify when we should do generic exception handling vs letting it propogate
3.  Disable the CA warning

Reply:
https://github.com/dotnet/upgrade-assistant/issues/813